crates/ruff_server/src/edit/document.rs

```diff
@@ -0,0 +1,120 @@
+use lsp_types::{Position, TextDocumentContentChangeEvent};
+use ruff_source_file::LineIndex;
+
+use crate::PositionEncoding;
+
+use super::range::text_range;
+
+#[derive(Debug, Clone)]
+pub struct Document {
+    contents: String,
+    index: LineIndex,
+    version: i32,
+}
+
+impl Document {
+    pub fn new(contents: String, version: i32) -> Self {
+        let index = LineIndex::from_source_text(&contents);
+        Self {
+            contents,
+            index,
+            version,
+        }
+    }
+    // TODO(jane): I would personally be in favor of removing access to this method and only
+    // allowing document mutation via specialized methods.
```

Comment:
I'm okay with this. The only use case that exists today is to replace the content. So maybe rename the method to `replace` in which case re-computing the `LineIndex` from scratch seems like a good default.